Fix frequent clipboard access in TextField by updating status

ThinkMars opened the pull request:
## Description

### Issue Summary
This PR fixes the issue where TextField was frequently calling `getPrimaryClipDescription` on Android, causing excessive clipboard access even when not needed.

### Root Cause Analysis
The `didUpdateWidget` method in `EditableTextState` was creating a circular dependency:
1. `didUpdateWidget` checks `pasteEnabled` condition
2. `pasteEnabled` depends on `clipboardStatus.value`
3. If conditions are met, `clipboardStatus.update()` is called
4. This triggers a rebuild, which calls `didUpdateWidget` again

### Benefits
- Reduces unnecessary platform calls to `getPrimaryClipDescription`
- Eliminates the rebuild loop
- Preserves all existing clipboard functionality
- Maintains focus handling and paste functionality unchanged

## Issue Link
Fixes #101196

## Files Modified
- `packages/flutter/lib/src/widgets/editable_text.dart`

## Testing
The fix has been verified to:
1. Stop excessive clipboard access
2. Maintain proper paste functionality
3. Work correctly with focus changes
4. Not break any existing tests

loic-sharma commented:
> The didUpdateWidget method in EditableTextState was creating a circular dependency:

@ThinkMars Sorry, I don't understand the circular dependency yet:

1. `didUpdateWidget` should only be called if the widget changes. How does calling  `clipboardStatus.update()` change the widget? (From what I can tell, changes to `clipboardStatus.value` calls `_onChangedClipboardStatus`, which calls `setState` but does not change the widget).
2. When `clipboardStatus.update()` is called, it won't call `_onChangedClipboardStatus` unless the status changed to a different value. For this to be the cause of an infinite loop, I would think that `clipboardStatus` would need to be constantly changing values. Is that so? If so, that seems highly unexpected - why is that happening?

By the way, this change will need a test. This is likely a good candidate to base your test off of: https://github.com/flutter/flutter/blob/0c98a23140421fdbff18b18c8487327382e5b01e/packages/flutter/test/material/text_field_test.dart#L15192-L15237

ThinkMars commented:
> getPrimaryClipDescription

@loic-sharma Sorry, my description may have caused a misunderstanding. Mainly due to our security check on Android phones, we found that frequent calls to the `getPrimaryClipDescription` method in the textField caused alarms.

Based on my debugPrint, I found that once operations such as inputting, selecting text, and pasting are performed within the textField, `getPrimaryClipDescription` is frequently called. This is considered too frequent by some security check tools and can be identified as a risk.

<img width="174" height="360" alt="image" src="https://github.com/user-attachments/assets/f341e766-b589-4d94-b6d1-52a52c30e916" />

<img width="324" height="198" alt="image" src="https://github.com/user-attachments/assets/579ff2d8-1126-4fb6-b946-0f48d38fec4d" />

After analysis, I found it was caused by frequent calls to the clipboard check. The specific method to trigger the check is `clipboardStatus.update()`. If can optimize the frequency of calling `hasStrings` in `clipboardStatus.update()`?

### Root Cause

1. **Frequent UI Reconstruction**: During user input, `TextEditingValue` changes cause the `EditableText`'s `build` method to execute frequently
2. **Condition Always True**: With default configuration, the condition `widget.selectionEnabled && pasteEnabled && canPaste` is always true
3. **No State Caching**: Each call to `clipboardStatus.update()` executes a system clipboard check without utilizing known status

### Solution

The solution I can think of is to add anti shake or throttling mechanisms, or provide some configurations to disable automatic checks triggered by the user's long press:

Firstly, add a throttle flag, it can be prevented to frequent or repeat to calls `hasStrings`.

<img width="542" height="438" alt="image" src="https://github.com/user-attachments/assets/2cf64ed1-58e8-4189-b940-344653221a42" />

Secondly, set the state to `unknown` in the `didChangeAppLifecycle` State cycle or `_copy` method so that it will only trigger in specific scenarios again.

ThinkMars committed:
fix(clipboard): Optimize clipboard status checking logic

Skip unnecessary checks when clipboard status is known
Force recheck status when app resumes to foreground or clipboard content changes

loic-sharma commented:
@ThinkMars I did some more more digging and discussed this problem with @justinmc.

It seems the root cause is that `EditableText` checks the status of the clipboard whenever the hover or focus state changes. So moving the pointer on/off or focusing/unfocusing the field all trigger a clipboard status update. This results in very frequent updates.

It seems the correct fix would be for the platform to "push" clipboard updates to the framework, instead of the framework "pulling" clipboard statuses from the platform. This would ensure Flutter's clipboard status is up-to-date even if some other background application updates the system clipboard. However, this is a major change as it would require updating all platforms Flutter supports.

Here's what might be a more pragmatic solution for today:

1. We re-add your changes to `didUpdateWidget` to update the clipboard only if its status is `ClipboardStatus.unknown`
2. We update `EditableTextState._handleFocusChanged` to update the clipboard if `_hasFocus` is true.

This would ensure that focusing the text field gets the latest status update, but hovering does not.

ThinkMars commented:
> @ThinkMars I did some more more digging and discussed this problem with @justinmc.
> 
> It seems the root cause is that `EditableText` checks the status of the clipboard whenever the hover or focus state changes. So moving the pointer on/off or focusing/unfocusing the field all trigger a clipboard status update. This results in very frequent updates.
> 
> It seems the correct fix would be for the platform to "push" clipboard updates to the framework, instead of the framework "pulling" clipboard statuses from the platform. This would ensure Flutter's clipboard status is up-to-date even if some other background application updates the system clipboard. However, this is a major change as it would require updating all platforms Flutter supports.
> 
> Here's what might be a more pragmatic solution for today:
> 
> 1. We re-add your changes to `didUpdateWidget` to update the clipboard only if its status is `ClipboardStatus.unknown`
> 2. We update `EditableTextState._handleFocusChanged` to update the clipboard if `_hasFocus` is true.
> 
> This would ensure that focusing the text field gets the latest status update, but hovering does not.

Okay，that may be a good solution. Thank you for your time! @loic-sharma

loic-sharma commented:
@ThinkMars, since this isn't ready for review yet, I'll go ahead and mark this as a draft. Please feel free to mark this as ready for review when you're ready. Thanks for the contribution!

ThinkMars committed:
Merge branch 'master' into fix/clipboard-status-update

Piinks commented:
Hey @ThinkMars I am going to close this PR for now as it has not been updated for quite a while. If you would like to add tests and re-open a fresh PR, please do! If you push a PR with the tests, they will run on CI if you are having issues locally. Thank you!
